lib/mail/delivery_recorder.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+module Mail
+  class DeliveryRecorder
+    def self.setup!(enabled:)
+      Mail::Message.include(MessageExtension)
+      ActionMailer::Base.register_observer(new(enabled: enabled))
+    end
+
+    def initialize(enabled:)
+      @enabled = enabled
+    end
+
+    def delivered_email(mail)
+      return unless enabled?
+
+      response = mail.delivery_method.response
+
+      ApplicationRecord.transaction do
+        email = Email.create!(
+          id: response.id,
+          from: response.content["from_email"],
+          to: mail.original_to,
+          template_id: response.template["id"],
+          template_version: response.template["version"],
+          uri: response.uri,
+          personalisation: mail.header["personalisation"].unparsed_value,
+        )
+
+        email.associate_with(*User.where(email: email.to).to_a)
+
+        mail.associations.each do |object, name|
+          email.associate_with object, as: name
+        end
+      end
+    end
+
+  private
+
+    def enabled?
+      !!@enabled
+    end
```

Comment:
What is this preventing?

Reply:
We (probably) don't want to record emails in test environment - we use TestDeliveryMethod so we cannot record all the fields. Additionally it would slow down the suite. If at any point we want to test the recording, there are number of tweaks to do this and I'd do it per-tests or per-context only.

Reply:
I mean the `!!` - is `if enabled?` ever different than `if @enabled`?

Reply:
I have a strong phobia of using instance variables in the code as I prefer separation between logic and data. Instance variable `@enabled` is pure data, `enabled?` is logic (which currently encapsulates the variable). That logic can change in the future and ensuring the code is using `enabled?` instead of `@enabled` minimises the amount of refactor required.